libs/guicore/pre/base/preprocessordataitem.h

```diff
@@ -21,15 +21,16 @@ class GUICOREDLL_EXPORT PreProcessorDataItem : public GraphicsWindowDataItem
 	PreProcessorDataItem(const QString& itemlabel, GraphicsWindowDataItem* parent);
 	PreProcessorDataItem(const QString& itemlabel, const QIcon& icon, GraphicsWindowDataItem* parent);
 	PreProcessorDataItem(ProjectDataItem* parent);
-	virtual ~PreProcessorDataItem();

```

Comment:
Hi Keisuke,

Shouldn't  this destructor still be virtual ?

Thanks,
Scott

Reply:
Hi Scott,

Thank you for the comment.

I've read in a book that when a member function (including destructor) is defined as a virtual function, inherited member functions in subclasses are automatically made virtual, so we do not need to add "virtual" keyword. That's why I've removed "virtual".

To make sure, I've written a sample program for testing.

```
#include <iostream>

class ClassWithoutVirtualDestructor1 {
public:
	ClassWithoutVirtualDestructor1();
	~ClassWithoutVirtualDestructor1();
};

ClassWithoutVirtualDestructor1::ClassWithoutVirtualDestructor1() {
	std::cout << "ClassWithoutVirtualDestructor1 constructor called" << std::endl;
}

ClassWithoutVirtualDestructor1::~ClassWithoutVirtualDestructor1() {
	std::cout << "ClassWithoutVirtualDestructor1 destructor called" << std::endl;
}

class ClassWithoutVirtualDestructor2 : public ClassWithoutVirtualDestructor1 {
public:
	ClassWithoutVirtualDestructor2();
	~ClassWithoutVirtualDestructor2();
};

ClassWithoutVirtualDestructor2::ClassWithoutVirtualDestructor2() {
	std::cout << "ClassWithoutVirtualDestructor2 constructor called" << std::endl;
}

ClassWithoutVirtualDestructor2::~ClassWithoutVirtualDestructor2() {
	std::cout << "ClassWithoutVirtualDestructor2 destructor called" << std::endl;
}

class ClassWithoutVirtualDestructor3 : public ClassWithoutVirtualDestructor2 {
public:
	ClassWithoutVirtualDestructor3();
	~ClassWithoutVirtualDestructor3();
};

ClassWithoutVirtualDestructor3::ClassWithoutVirtualDestructor3() {
	std::cout << "ClassWithoutVirtualDestructor3 constructor called" << std::endl;
}

ClassWithoutVirtualDestructor3::~ClassWithoutVirtualDestructor3() {
	std::cout << "ClassWithoutVirtualDestructor3 destructor called" << std::endl;
}

class ClassWithVirtualDestructor1 {
public:
	ClassWithVirtualDestructor1();
	virtual ~ClassWithVirtualDestructor1();
};

ClassWithVirtualDestructor1::ClassWithVirtualDestructor1() {
	std::cout << "ClassWithVirtualDestructor1 constructor called" << std::endl;
}

ClassWithVirtualDestructor1::~ClassWithVirtualDestructor1() {
	std::cout << "ClassWithVirtualDestructor1 destructor called" << std::endl;
}

class ClassWithVirtualDestructor2 : public ClassWithVirtualDestructor1 {
public:
	ClassWithVirtualDestructor2();
	~ClassWithVirtualDestructor2();
};

ClassWithVirtualDestructor2::ClassWithVirtualDestructor2() {
	std::cout << "ClassWithVirtualDestructor2 constructor called" << std::endl;
}
ClassWithVirtualDestructor2::~ClassWithVirtualDestructor2() {
	std::cout << "ClassWithVirtualDestructor2 destructor called" << std::endl;
}

class ClassWithVirtualDestructor3 : public ClassWithVirtualDestructor2 {
public:
	ClassWithVirtualDestructor3();
	~ClassWithVirtualDestructor3();
};

ClassWithVirtualDestructor3::ClassWithVirtualDestructor3() {
	std::cout << "ClassWithVirtualDestructor3 constructor called" << std::endl;
}

ClassWithVirtualDestructor3::~ClassWithVirtualDestructor3() {
	std::cout << "ClassWithVirtualDestructor3 destructor called" << std::endl;
}

void testClassesWithoutVirtualDestructor()
{
	std::cout << "Entered testClassesWithoutVirtualDestructor()" << std::endl;

	// ClassWithoutVirtualDestructor3 created as instance and deleted
	// constructors and destructors of three classes are correctly called
	{
		std::cout << "Creating and deleting ClassWithoutVirtualDestructor3 normally" << std::endl;
		ClassWithoutVirtualDestructor3 obj;
	}

	// ClassWithoutVirtualDestructor3 created through ClassWithoutVirtualDestructor3* and deleted
	// constructors and destructors of three classes are correctly called
	{
		std::cout << "Creating and deleting ClassWithoutVirtualDestructor3 through ClassWithoutVirtualDestructor3*" << std::endl;
		ClassWithoutVirtualDestructor3* obj = new ClassWithoutVirtualDestructor3();
		delete obj;
	}

	// ClassWithoutVirtualDestructor3 created through ClassWithoutVirtualDestructor1* and deleted
	// constructors of three classes are correctly called, but destructors of ClassWithoutVirtualDestructor2, ClassWithoutVirtualDestructor3 are not called.
	{
		std::cout << "Creating and deleting ClassWithoutVirtualDestructor3 through ClassWithoutVirtualDestructor1*" << std::endl;
		ClassWithoutVirtualDestructor1* obj = new ClassWithoutVirtualDestructor3();
		delete obj;
	}
	// ClassWithoutVirtualDestructor3 created through ClassWithoutVirtualDestructor2* and deleted
	// constructors of three classes are correctly called, but destructors of ClassWithoutVirtualDestructor3 is not called.
	{
		std::cout << "Creating and deleting ClassWithoutVirtualDestructor3 through ClassWithoutVirtualDestructor2*" << std::endl;
		ClassWithoutVirtualDestructor2* obj = new ClassWithoutVirtualDestructor3();
		delete obj;
	}
}

void testClassesWithVirtualDestructor()
{
	std::cout << "Entered testClassesWithVirtualDestructor()" << std::endl;

	// ClassWithVirtualDestructor3 created as instance and deleted
	// constructors and destructors of three classes are correctly called
	{
		std::cout << "Creating and deleting ClassWithVirtualDestructor3 normally" << std::endl;
		ClassWithVirtualDestructor3 obj;
	}

	// ClassWithoutVirtualDestructor3 created through ClassWithoutVirtualDestructor3* and deleted
	// constructors and destructors of three classes are correctly called
	{
		std::cout << "Creating and deleting ClassWithVirtualDestructor3 through ClassWithVirtualDestructor3*" << std::endl;
		ClassWithVirtualDestructor3* obj = new ClassWithVirtualDestructor3();
		delete obj;
	}

	// ClassWithoutVirtualDestructor3 created through ClassWithoutVirtualDestructor1* and deleted
	// constructors and destructors of three classes are correctly called
	{
		std::cout << "Creating and deleting ClassWithVirtualDestructor3 through ClassWithVirtualDestructor1*" << std::endl;
		ClassWithVirtualDestructor1* obj = new ClassWithVirtualDestructor3();
		delete obj;
	}
	// ClassWithoutVirtualDestructor3 created through ClassWithoutVirtualDestructor2* and deleted
	// constructors and destructors of three classes are correctly called
	{
		std::cout << "Creating and deleting ClassWithVirtualDestructor3 through ClassWithVirtualDestructor2*" << std::endl;
		ClassWithVirtualDestructor2* obj = new ClassWithVirtualDestructor3();
		delete obj;
	}
}

int main()
{
	testClassesWithoutVirtualDestructor();
	testClassesWithVirtualDestructor();
	
	return 0;
}
```

You can save it as test.cpp, and build with `cl test.cpp`.

I've built with Visual C++ 2013, and got the result below.


```
Entered testClassesWithoutVirtualDestructor()
Creating and deleting ClassWithoutVirtualDestructor3 normally
ClassWithoutVirtualDestructor1 constructor called
ClassWithoutVirtualDestructor2 constructor called
ClassWithoutVirtualDestructor3 constructor called
ClassWithoutVirtualDestructor3 destructor called
ClassWithoutVirtualDestructor2 destructor called
ClassWithoutVirtualDestructor1 destructor called
Creating and deleting ClassWithoutVirtualDestructor3 through ClassWithoutVirtualDestructor3*
ClassWithoutVirtualDestructor1 constructor called
ClassWithoutVirtualDestructor2 constructor called
ClassWithoutVirtualDestructor3 constructor called
ClassWithoutVirtualDestructor3 destructor called
ClassWithoutVirtualDestructor2 destructor called
ClassWithoutVirtualDestructor1 destructor called
Creating and deleting ClassWithoutVirtualDestructor3 through ClassWithoutVirtualDestructor1*
ClassWithoutVirtualDestructor1 constructor called
ClassWithoutVirtualDestructor2 constructor called
ClassWithoutVirtualDestructor3 constructor called
ClassWithoutVirtualDestructor1 destructor called
Creating and deleting ClassWithoutVirtualDestructor3 through ClassWithoutVirtualDestructor2*
ClassWithoutVirtualDestructor1 constructor called
ClassWithoutVirtualDestructor2 constructor called
ClassWithoutVirtualDestructor3 constructor called
ClassWithoutVirtualDestructor2 destructor called
ClassWithoutVirtualDestructor1 destructor called
Entered testClassesWithVirtualDestructor()
Creating and deleting ClassWithVirtualDestructor3 normally
ClassWithVirtualDestructor1 constructor called
ClassWithVirtualDestructor2 constructor called
ClassWithVirtualDestructor3 constructor called
ClassWithVirtualDestructor3 destructor called
ClassWithVirtualDestructor2 destructor called
ClassWithVirtualDestructor1 destructor called
Creating and deleting ClassWithVirtualDestructor3 through ClassWithVirtualDestructor3*
ClassWithVirtualDestructor1 constructor called
ClassWithVirtualDestructor2 constructor called
ClassWithVirtualDestructor3 constructor called
ClassWithVirtualDestructor3 destructor called
ClassWithVirtualDestructor2 destructor called
ClassWithVirtualDestructor1 destructor called
Creating and deleting ClassWithVirtualDestructor3 through ClassWithVirtualDestructor1*
ClassWithVirtualDestructor1 constructor called
ClassWithVirtualDestructor2 constructor called
ClassWithVirtualDestructor3 constructor called
ClassWithVirtualDestructor3 destructor called
ClassWithVirtualDestructor2 destructor called
ClassWithVirtualDestructor1 destructor called
Creating and deleting ClassWithVirtualDestructor3 through ClassWithVirtualDestructor2*
ClassWithVirtualDestructor1 constructor called
ClassWithVirtualDestructor2 constructor called
ClassWithVirtualDestructor3 constructor called
ClassWithVirtualDestructor3 destructor called
ClassWithVirtualDestructor2 destructor called
ClassWithVirtualDestructor1 destructor called
```

In this case, destructor of ClassWithVirtualDestructor2 and ClassWithVirtualDestructor3 are not defined with "virtual" keyword, but in testClassesWithVirtualDestructor(), destructors of three classes are correctly called in all cases.

In case of PreProcessorDataItem, it is a subclass of QObject, that defines virtual destructor, so I've noticed that I can remove "virtual" keyword from its destructor.

Anyway, I'm sorry that it was a confusing modification. Actually, in the source code of iRIC, most of destructors of subclasses are defined with "virtual" keyword, because I've written most of the source code before I read the book.

I think I should have fixed destructors of all subclasses at once, but in this case I've modified only one file that I've noticed. Anyway, please understand that the code works correctly.

Thank you,

Keisuke Inoue

Reply:
Hi Keisuke,

Thanks for that information, I didn't know that.  I thought that removing the `virtual` keyword would make it non-virtual.  Through a little research I found out that you can do this using the `final` keyword (similar to java).  Also, I found that it is recommended to use the `overrides` keyword for virtual inheritance [(even for destructors)](https://github.com/Microsoft/clang-tools-extra/blob/master/test/clang-tidy/modernize-use-override.cpp#L48).  I think it is useful since it shows that the dtor is virtual and will also produce an error if the base class dtor isn't virtual.

Thanks,
Scott

Reply:
Hi Scott,

Thank you for comment and suggestion. I agree that it is better to add "override" keyword. It will help us notice that it actually is a inherited virtual function.

I've pushed another commit to remove "virtual" and add "override" for member functions of PreProcessorDataItem.

I've also registered #818 so that it will remind that we should make similar modifications to other classes. I will work on this when I've found time.

Thank you,

Keisuke